### PR TITLE
Changed TOPI ops to use C++ schedules

### PR DESCRIPTION
**Pull request by alex-weaver**
Added a ScheduleFactory class to dispatch target-specific schedules, and TOPI ops now use this for scheduling rather than the python schedules.
#343

### CODE REVIEW

**Comment by tqchen** [comment]
Let us spend more time to think about this before getting to a c++ api. ScheduleFactory is fine for c++ alone, but we have additional requirements for being able to interacting with the python side and interact with the existing generic dispatching module.

**Comment by alex-weaver** [comment]
Ok that makes sense. If the python and C++ need to cooperate, then does this mean that the C++ should maintain the registry and should expose a way for python to register specializations?

**Comment by clouds56** [comment]
Is there any update or anything I could here?

**Comment by alex-weaver** [comment]
@clouds56 This PR is currently mostly blocked on the TVM PR https://github.com/dmlc/tvm/pull/892
once that one is resolved, I think there should only be one further obstacle to this PR working: in the current implementation of https://github.com/dmlc/tvm/pull/892 the C++ and Python schedules are completely separate, although the capability certainly exists for them to interoperate. Figuring out the details of that will probably be the main obstacle for this PR. Having some way for the C++ schedules to be registered by default, and then allow the python ones to override them is the main aim with that.

**Comment by alex-weaver** [comment]
@tqchen Ok this is now updated to use the new C++ GenericFunc system in TVM for schedules

File: include/nnvm/top/util.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#include <tvm/build_module.h>
+#include <nnvm/base.h>
+#include <nnvm/compiler/op_attr_types.h>
+
+#include <string>
+
+namespace nnvm {
+namespace top {
+
+inline nnvm::compiler::FTVMSchedule MakeScheduleQuery(const std::string& name) {
```
Comment (tqchen) [inline include/nnvm/top/util.h]: document what this is about.

**Pull request closed without merging**